File: src/main/java/net/frozenorb/potpvp/util/menu/pagination/PaginatedMenu.java
package net.frozenorb.potpvp.util.menu.pagination;

import net.frozenorb.potpvp.util.menu.Button;
import net.frozenorb.potpvp.util.menu.Menu;
import lombok.Getter;
import org.bukkit.entity.Player;

import java.util.HashMap;
import java.util.Map;

public abstract class PaginatedMenu extends Menu {

    @Getter private int page = 1;

    @Override
    public String getTitle(Player player) {
        return getPrePaginatedTitle(player) + " - " + page + "/" + getPages(player);
    }

    /**
     * Changes the page number
     *
     * @param player player viewing the inventory
     * @param mod    delta to modify the page number by
     */
    public final void modPage(Player player, int mod) {
        this.page += mod;

        this.getButtons().clear();
        this.openMenu(player);
    }

    /**
     * @param player player viewing the inventory
     * @return
     */
    public final int getPages(Player player) {

        final int buttonAmount = this.getAllPagesButtons(player).size();

        return buttonAmount == 0 ? 1 : (int)Math.ceil((double)buttonAmount / (double)this.getMaxItemsPerPage(player));
    }

    @Override
    public final Map<Integer, Button> getButtons(Player player) {

        final int minIndex = (int) ((double) (page - 1) * getMaxItemsPerPage(player));
        final int maxIndex = (int) ((double) (page) * getMaxItemsPerPage(player));

        final HashMap<Integer, Button> buttons = new HashMap<>();

        buttons.put(0, new PageButton(-1, this));
        buttons.put(8, new PageButton(1, this));

        for (Map.Entry<Integer, Button> entry : getAllPagesButtons(player).entrySet()) {

            int ind = entry.getKey();

            if (ind >= minIndex && ind < maxIndex) {

                ind -= (int) ((double) (getMaxItemsPerPage(player)) * (page - 1)) - 9;
                buttons.put(ind, entry.getValue());
            }

        }

        final Map<Integer, Button> global = getGlobalButtons(player);

        if (global != null) {

            for (Map.Entry<Integer, Button> gent : global.entrySet()) {
                buttons.put(gent.getKey(), gent.getValue());
            }

        }


        return buttons;
    }

    public int getMaxItemsPerPage(Player player) {
        return 18;
    }

    /**
     * @param player player viewing the inventory
     * @return a Map of buttons that returns items which will be present on all pages
     */
    public Map<Integer, Button> getGlobalButtons(Player player) {
        return null;
    }

    /**
     * @param player player viewing the inventory
     * @return title of the inventory before the page number is added
     */
    public abstract String getPrePaginatedTitle(Player player);

    /**
     * @param player player viewing the inventory
     * @return a map of buttons that will be paginated and spread across pages
     */
    public abstract Map<Integer, Button> getAllPagesButtons(Player player);
}

File: src/main/kotlin/net/frozenorb/potpvp/kt/menu/pagination/PaginatedMenu.kt
package net.frozenorb.potpvp.kt.menu.pagination

import net.frozenorb.potpvp.kt.menu.Button
import net.frozenorb.potpvp.kt.menu.Menu
import org.bukkit.entity.Player
import java.util.HashMap

abstract class PaginatedMenu : Menu() {

    internal var page: Int = 1

    override fun getTitle(player: Player): String {
        return getPrePaginatedTitle(player) + " - " + page + "/" + getPages(player)
    }

    fun modPage(player: Player, mod: Int) {
        page += mod
        buttons.clear()
        openMenu(player)
    }

    internal fun getPages(player: Player): Int {
        val buttonAmount = getAllPagesButtons(player).size
        return if (buttonAmount == 0) {
            1
        } else Math.ceil(buttonAmount / getMaxItemsPerPage(player).toDouble()).toInt()
    }

    override fun getButtons(player: Player): MutableMap<Int, Button> {
        val minIndex = ((page - 1) * getMaxItemsPerPage(player).toDouble()).toInt()
        val maxIndex = (page * getMaxItemsPerPage(player).toDouble()).toInt()
        val buttons = HashMap<Int, Button>()
        buttons[0] = PageButton(-1, this)
        buttons[8] = PageButton(1, this)

        for (entry in getAllPagesButtons(player).entries) {
            var ind = entry.key
            if (ind in minIndex until maxIndex) {
                ind -= (getMaxItemsPerPage(player) * (page - 1).toDouble()).toInt() - 9
                buttons[ind] = entry.value
            }
        }

        val global = getGlobalButtons(player)
        if (global != null) {
            for ((key, value) in global) {
                buttons[key] = value
            }
        }

        return buttons
    }

    open fun getMaxItemsPerPage(player: Player): Int {
        return 18
    }

    open fun getGlobalButtons(player: Player): Map<Int, Button>? {
        return null
    }

    abstract fun getPrePaginatedTitle(player: Player): String

    abstract fun getAllPagesButtons(player: Player): Map<Int, Button>

}